crypto/keys/internal/ecdsa/privkey.go

```diff
@@ -38,10 +77,35 @@ func (sk *PrivKey) Bytes() []byte {
 	return bz
 }
 
-// Sign hashes and signs the message usign ECDSA. Implements SDK PrivKey interface.
+// Sign hashes and signs the message using ECDSA. Implements SDK
+// PrivKey interface.
+// NOTE: this now calls the ecdsa Sign function
+// (not method!) directly as the s value of the signature is needed to
+// low-s normalize the signature value
+// It then ASN DER encodes in exactly the same way as the underlying
+// go-lang crypto code does (as of 7/20/2021 anyway)
+
 func (sk *PrivKey) Sign(msg []byte) ([]byte, error) {
+
 	digest := sha256.Sum256(msg)
-	return sk.PrivateKey.Sign(rand.Reader, digest[:], nil)
+	r, s, err := ecdsa.Sign(rand.Reader, &sk.PrivateKey, digest[:])
+
+	if err != nil {
+
+		return nil, err
+
+	}
+
+	normS := NormalizeS(s)
+	var b cryptobyte.Builder
+
+	b.AddASN1(asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+
+		b.AddASN1BigInt(r)
+		b.AddASN1BigInt(normS)
+	})
```

Comment:
If at all possible, I'd strongly suggest using fixed-sized ECDSA signatures here.

The ASN.1 DER serialization just adds incidental complexity, almost always resulting in larger signatures than the fixed-width encoding.

It's also more parsimonious with the existing ECDSA/secp256k1 signatures, and could permit more code reuse across curves.

Reply:
Thanks Tony. 

Is this signature serialization documented somewhere? bip62 shows a DER encoding. If someone implemented cosmos signature verification in some other language than go-lang (or even with unmodified golang crypto ecdsa) they wouldn't be able to verify these signatures would they? 

I'd also suggest that this PR (for signature malleability) is probably not the right place for this change.

Reply:
The best spec I'm aware of for fixed-sized ECDSA signatures is unfortunately PKCS#11, which is a spec I wouldn't wish on anyone.

However, it's the same signature format that ECDSA/secp256k1 signatures in Cosmos SDK/Tendermint are already using.

I'm just suggesting that the signature format is consistent regardless of the elliptic curve used for ECDSA (and that, in general, the fixed-sized format is much simpler)

Reply:
Yeah, I hear you. In general, I think the secp256r1 code is quite different than the secp256k1 code, and I think there is a lot that could be factored out into a general ECDSA suite for blockchain signatures that includes both curves (and perhaps allows secp curves with more bits to reuse code too). 

Unfortunately, I'm quite familiar with PKCS11 already, and I know what you mean :)

Reply:
So, I looked at btcec, which is the golang library for secp256k1. That implements bip62 (https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#DER_encoding) which uses DER - https://github.com/btcsuite/btcd/blob/63438c6d3661a7dec09013f7731035abdf1eb4bc/btcec/signature.go#L63

If we want to align with the secp256k1 implementation, we should do the same. 

That would invalidate (unless we additionally check for, parse and validate that encoding too) existing secp256r1 go-lang/DER-formatted signatures. I don't know if that's an issue, but could be.

I don't see a strong reason to implement anything other than bip62 DER, or continue (for now) to use the existing "normal" go-lang DER in this PR. The only reason I can see to use bip62 encoding is for consistency with the secp256k1 code. Since they are currently separate code paths, I don't even see that as a pressing need. 

As an aside though, I do think it's worth creating a new issue to suggest refactor the Cosmos signature code so we can effectively align ECDSA signatures using any curve without having to implement the same code in multiple places. That might include, for example (strawman), making a contribution to btcec implementing secp256r1, and then reusing btcec in the cosmos code for all ECDSA curves supported by btcec.

Reply:
> If we want to align with the secp256k1 implementation, we should do the same.

Tendermint/Cosmos SDK use fixed-width secp256k1 signatures. If you want to align the two implementations, it should use fixed width secp256r1/P-256 signatures as well.

Using DER at all was something of a historical accident stemming from Bitcoin, and most newer cryptocurrency systems have opted to use fixed-width signatures instead due to all of the issues/complexity involved with DER.

Reply:
Here's the existing secp256k1 serialization code in Cosmos SDK, for example:

https://github.com/cosmos/cosmos-sdk/blob/320a852/crypto/keys/secp256k1/secp256k1_nocgo.go#L60-L70

Reply:
I just pushed new changes that implement the raw encoding of the signature, using the secp256k1 function unchanged (except to deal with the slightly different context in secp256r1 code). So I'm going to resolve this thread.